spark/src/main/java/org/apache/iceberg/spark/actions/BaseRewriteDataFilesSparkAction.java

```diff
@@ -0,0 +1,443 @@
+        Thread.yield();
+        // Either we have a full commit group, or we have completed writing and need to commit what is left over
+        if (completedRewriteIds.size() > groupsPerCommit ||
+            (!stillRewriting.get() && completedRewriteIds.size() > 0)) {
+
+          Set<String> batch = Sets.newHashSetWithExpectedSize(groupsPerCommit);
+          for (int i = 0; i < groupsPerCommit && !completedRewriteIds.isEmpty(); i++) {
+            batch.add(completedRewriteIds.poll());
+          }
+
+          try {
+            commitOrClean(batch);
+            completedCommitIds.addAll(batch);
+          } catch (Exception e) {
+            batch.forEach(results::remove);
+            LOG.error("Failure during rewrite commit process, partial progress enabled. Ignoring", e);
+          }
+        }
+      }
+    });
+
+    // Start rewrite tasks
+    Tasks.foreach(groupStream.iterator())
+        .suppressFailureWhenFinished()
+        .executeWith(rewriteService)
+        .noRetry()
+        .onFailure((info, exception) -> {
+          LOG.error("Failure during rewrite process for group {}", info.first(), exception);
+          abortFileGroup(info.first().groupID);
+        })
+        .run(infoListPair ->
+            rewriteFiles(infoListPair, totalGroups, numGroupsPerPartition, strategy, completedRewriteIds, results));
+
+    stillRewriting.set(false);
+    committerService.shutdown();
+
+    try {
+      committerService.awaitTermination(10, TimeUnit.MINUTES);
+    } catch (InterruptedException e) {
+      throw new RuntimeException("Cannot complete commit for rewrite, commit service interuptted", e);
+    }
+
+    return new Result(results.values().stream().collect(Collectors.toMap(Pair::first, Pair::second)));
+  }
+
+  @Override
+  public Result execute() {
+    RewriteStrategy strategy =  rewriteStrategy(strategyType).options(this.options());
+    validateOptions(strategy);
+
+    CloseableIterable<FileScanTask> files = files();
+
+    Map<StructLike, List<List<FileScanTask>>> fileGroupsByPartition = filterAndGroupFiles(files, strategy);
+
+    try {
+      files.close();
+    } catch (IOException io) {
+      LOG.error("Cannot properly close file iterable while planning for rewrite", io);
+    }
+
+    Map<StructLike, Integer> numGroupsPerPartition = fileGroupsByPartition.entrySet().stream()
+        .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().size())
+    );
+
+    Integer totalGroups = numGroupsPerPartition.values().stream().reduce(Integer::sum).orElse(0);
+
+    if (totalGroups == 0) {
+      return new Result(Collections.emptyMap());
+    }
+
+    Map<StructLike, Integer> partitionIndex = new HashMap<>();
+    AtomicInteger jobIndex = new AtomicInteger(1);
+
+    // Todo Check if we need to randomize the order in which we do jobs, instead of being partition centric
+    Stream<Pair<FileGroupInfo, List<FileScanTask>>> groupStream = fileGroupsByPartition.entrySet().stream()
```

Comment:
I think by creating `toJobStream(fileGroupsByPartition)`, you should be able encapsulate the index vars above.

Reply:
Also moving execute up to be with the other public overrides